src/github.com/getlantern/balancer/strategy.go

```diff
@@ -0,0 +1,106 @@
+package balancer
+
+import (
+	"math/rand"
+)
+
+// Random strategy gives even chance to each dialer, act as a baseline to other
+// strategies.
+func Random(dialers []*dialer) dialerHeap {
+	return dialerHeap{dialers, func(i, j int) bool {
+		// we don't need good randomness, skip seeding
+		if rand.Intn(2) == 0 {
+			return false
+		}
+		return true
+	}}
+}
+
+// Sticky strategy always pick the dialer with largest consecutive success
+// count or the smallest consecutive failure count
+func Sticky(dialers []*dialer) dialerHeap {
+	return dialerHeap{dialers, func(i, j int) bool {
+		mi := dialers[i].metrics()
+		mj := dialers[j].metrics()
+		return (mi.consecSuccesses - mi.consecFailures) >
+			(mj.consecSuccesses - mj.consecFailures)
+	}}
+}
+
+// Fastest strategy always pick the dialer with lowest recent average connect time
+func Fastest(dialers []*dialer) dialerHeap {
+	return dialerHeap{dialers, func(i, j int) bool {
+		mi := dialers[i].metrics()
+		mj := dialers[j].metrics()
+		return mi.avgConnTime < mj.avgConnTime
```

Comment:
Thinking twice about throughput, there are 3 main physical factors: bandwidth, latency and packet drop rate. In normal condition, TCP flow control will gradually reach the limit of lowest bandwidth link, which is usually user's local network.

Apart from latency, the main factor to differentiate throughput among proxies is network congestion at international trunk or the network proxies reside, which is roughly random to TCP endpoints, so TCP flow control can't handle it gracefully, hence results more dropped packets. Packet dropping has two consequences:

1. Increase the time required to `Read()` or connect, because of TCP retransmission.

2. Fail to connect or close prematurely, if the congestion is serious enough.

Consider the randomness of network congestion, the packet drop rate should be the same at each stage of a TCP connection, so statistically, the delay and fail rate to establish connections can represent the overall delay and fail rate, unless there's routing policy somewhere to drop more larger packets than smaller ones.

That is, we don't need to measure throughput separately. Or, if we confirmed that drop rate for large packets is higher, just periodically check with large enough payload for single packet.

Is it right @oxtoacart ?

Inspired by:
1. https://blog.thousandeyes.com/a-very-simple-model-for-tcp-throughput/
2. http://bradhedlund.com/2008/12/19/how-to-calculate-tcp-throughput-for-long-distance-links/